### PR TITLE
tests: add mz e2e tests

### PR DESCRIPTION
**Pull request by joacoc**
### Motivation

This PR intends to implement the E2E tests for `mz`. A long missing issue to release the CLI. https://github.com/MaterializeInc/materialize/issues/21168

### Tips for reviewer

My first time building tests in this framework, so expect rookie things.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by joacoc** [comment]
@MaterializeInc/testing can I test this in buildkite before merging?

**Comment by nrainer-materialize** [comment]
> @MaterializeInc/testing can I test this in buildkite before merging?

Is your question if you can run Nightly for this pull request? It is automatically triggered because the Nightly definition file was changed.

**Comment by joacoc** [comment]
@nrainer-materialize where is it? I cant find it in the buildkite test maze. Maybe I misconfigured something.

**Comment by joacoc** [comment]
@nrainer-materialize - `MZ_CLI_APP_PASSWORD` is stored as [Pulumi secrets in the i2 repository.](https://github.com/MaterializeInc/i2/pull/1105)

Why could it be that is missing in Buildkite?

```
APP_PASSWORD = os.environ["MZ_CLI_APP_PASSWORD"]
--
  | ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  | File "<frozen os>", line 679, in __getitem__
  | KeyError: 'MZ_CLI_APP_PASSWORD'
```

File: test/mz-e2e/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    _wait_for_pg,
+)
+from materialize.mzcompose.composition import (
+    Composition,
+    WorkflowArgumentParser,
+)
+from materialize.mzcompose.services.cockroach import Cockroach
+from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.mz import Mz
+from materialize.mzcompose.services.testdrive import Testdrive
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:30]: You do not use Testdrive at all, so consider removing it here and all mentions below.

File: test/mz-e2e/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        environment=ENVIRONMENT,
+        app_password=APP_PASSWORD,
+    ),
+]
+
+
+def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+    """Deploy the current source to the cloud and run tests."""
+
+    parser.add_argument(
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:65]: I do not think you are using any of those command-line arguments, so consider removing them.

File: test/mz-e2e/mzcompose.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        except UIError:
+            # Work around https://github.com/MaterializeInc/materialize/issues/17219
+            pass
+
+        time.sleep(10)
+
+        assert "materialize.cloud" in cloud_hostname(c)
+        wait_for_cloud(c)
+
+        # if args.version_check:
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:104]: I think there is no point in doing a version_check in this workflow, as the cloud-canary workflow will do it and will report any problems. So consider removing the entire machinery associated with version checks -- methods, command-line arguments.

File: test/mz-e2e/mzcompose.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        #     version_check(c)
+
+        # Test - `mz app-password`
+        # Assert `mz app-password create`
+        new_app_password_name = "Materialize CLI (mz) - Nightlies"
+        output = c.run(
+            "mz", "app-password", "create", new_app_password_name, capture=True
+        )
+        assert "mzp_" in output.stdout
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:113]: You do not check that the password thus obtained can actually be used to connect to the Materialize instance.

File: test/mz-e2e/mzcompose.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        # Test - `mz region list`
+        # Enable, disable and show are already tested.
+        output = c.run("mz", "region", "list", "--format", "json", capture=True)
+        regions = json.loads(output.stdout)
+        us_region = regions[0]
+        if us_region["region"] != "aws/us-east-1":
+            us_region = regions[1]
+        assert "enabled" == us_region["status"]
+
+        # Look for the '.psqlrc' file in the home dir.
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:178]: Which command is supposed to create this file? Consider making sure the file does not exist to start with, before running the command.

File: test/mz-e2e/mzcompose.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        home_dir = os.path.expanduser("~")
+        path = os.path.join(home_dir, ".psqlrc-mz")
+
+        # Verify the content is ok
+        if os.path.exists(path):
+            with open(path) as file:
+                content = file.read()
+
+            if content != "\\timing\n\\include ~/.psqlrc":
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:187]: Ideally, one should use the actual `psql` client to determine that the `.psql-mz` file is readable by the tool that is meant to read it. We export a psql container that can be used to run psql.

File: test/mz-e2e/mzcompose.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        assert new_app_password_name in output.stdout
+
+        # // Test - `mz secrets`
+        secret = "decode('c2VjcmV0Cg==', 'base64')"
+        output = c.run(
+            "mz", "secret", "create", "CI_SECRET", stdin=secret, capture=True
+        )
+        assert output.returncode == 0
+
+        output = c.run("mz", "sql", "--", "-q", "-c", "SHOW SECRETS;", capture=True)
```
Comment (philip-stoev) [inline test/mz-e2e/mzcompose.py:126]: We have had regressions in the past where a secret would appear in `SHOW SECRETS` but then could not actually be used to create a database object. So I suggest you create a secret that actually exists in the real world and use it to connect to a real source. In Buildkite, `CONFLUENT_API_KEY` and `CONFLUENT_API_SECRET` exist and you can use them to connect to an actual Kafka source -- see `test/cloud-canary/canary-sources.td` for an example.

**Review by philip-stoev — CHANGES_REQUESTED** [review]
Thank you very much for the test and apologies if I left one too many comments.

Please let me know if one thing or another is difficult to test.

**Comment by joacoc** [comment]
Thank you, @philip-stoev. I will fix them now. Do you have any idea why the env var `MZ_CLI_APP_PASSWORD` is not available in Buildkite?

**Comment by philip-stoev** [comment]
> Thank you, @philip-stoev. I will fix them now. Do you have any idea why the env var `MZ_CLI_APP_PASSWORD` is not available in Buildkite?

In order for a variable to be available in Buildkite it needs to be declared in the following sections:

1. In the `i2` repository:
 - Pulumi.produciton.yaml
 - i2/buildkite.py, line 325

2. In the main materialize repository:
 - in bin/ci-builder line 161

File: test/mz-e2e/mzcompose.py (hunk 6 above)
Comment (joacoc) [reply test/mz-e2e/mzcompose.py:187]: `mz sql` executes `psql` under the hood to use file. This is just an extra step to assert the file doesn't has anything additional

**Review by philip-stoev — APPROVED** [review]
Thank you!

**Pull request merged**